Refactor `Commands` module

neonichu opened the pull request:
This is the initial step in reducing the first stage of the bootstrap build. Currently, the initial bootstrap build using CMake is essentially building the entirety of SwiftPM and that is (mostly) because the `Commands` module pulls in everything that SwiftPM possibly offers.

This introduces a new `CoreCommands` module which contains the necessary parts for providing a minimal build command using the native build system. This can be used to provide a new `swift-bootstrap` executable which is a minimal version of `swift-build`, but actually doing so will be a separate PR.

The Windows build currently relies entirely on CMake, so initally we will not be able to profit much from the minimal build, but this will get us closer to doing so in the future.

tomerd commented:
very cool!

neonichu commented:
@swift-ci please smoke test

neonichu commented:
https://github.com/apple/sourcekit-lsp/pull/650
@swift-ci please smoke test

neonichu commented:
https://github.com/apple/swift/pull/61574
https://github.com/apple/sourcekit-lsp/pull/650
@swift-ci please smoke test

stevapple commented:
Bootstrapping SwiftPM in Windows with "debug" configuration is possible, but in release build it will hit apple/swift-llbuild#844, so this is currently unusable. It’s nice to see SwiftPM being bootstrapped on Windows, in align with other platforms.

neonichu commented:
https://github.com/apple/swift/pull/61574
https://github.com/apple/sourcekit-lsp/pull/650
@swift-ci please smoke test macOS

neonichu commented:
`fatal error encountered while reading from module 'CxxVectorSum'`

Looks like this is an unrelated issue on the CI right now? I also saw it on TSC PRs.

neonichu commented:
https://github.com/apple/swift/pull/61574
@swift-ci please smoke test macOS

neonichu commented:
Split out #5846 to address the concerns about build system support.

neonichu commented:
Simplified this PR a little bit for easier review and updated it based on the refactoring I have done separately.

Once this lands, there will be two eventual follow-up PRs:
- a pretty simple PR that introduces the `swift-bootstrap` command and uses it for the `bootstrap` script
- a PR which changes the bootstrap process and CMake configuration to actually be more minimal. This requires us to move the Windows build to do bootstrapping first, so it'll take us a few steps to get there.

neonichu commented:
@swift-ci please smoke test

neonichu commented:
@swift-ci please smoke test

tomerd commented on Sources/Commands/SwiftBuildTool.swift:
```diff
@@ -157,8 +159,23 @@ public struct SwiftBuildTool: SwiftCommand {
+        return .init(providers: try swiftTool.defaultBuildSystemProvider.providers.merging([
+            .xcode: { (explicitProduct: String?, cacheBuildManifest: Bool, customBuildParameters: BuildParameters?, customPackageGraphLoader: (() throws -> PackageGraph)?, customOutputStream: OutputByteStream?, customLogLevel: Basics.Diagnostic.Severity?, customObservabilityScope: ObservabilityScope?) throws -> BuildSystem in
+                let graphLoader = { try swiftTool.loadPackageGraph(explicitProduct: explicitProduct) }
+                // FIXME: Implement the custom build command provider also.
```
still to do?

neonichu replied:
I just carried this over from the original code, honestly not sure what it means, so maybe we can remove it?

neonichu replied:
Removing this now.

tomerd commented on Sources/Commands/SwiftPackageTool.swift:
```diff
@@ -1190,7 +1191,7 @@ final class PluginDelegate: PluginInvocationDelegate {
 
         // Configure the verbosity of the output.
-        let logLevel: Diagnostic.Severity
+        let logLevel: Basics.Diagnostic.Severity
```
needed?

neonichu replied:
Yep

tomerd commented on Sources/Commands/Utilities/APIDigester.swift:
```diff
@@ -62,7 +63,7 @@ struct APIDigesterBaselineDumper {
         at baselineDir: AbsolutePath?,
         force: Bool,
-        logLevel: Diagnostic.Severity,
+        logLevel: Basics.Diagnostic.Severity,
```
needed?

neonichu replied:
Yep

tomerd commented on Sources/Commands/SwiftBuildTool.swift:
```diff
@@ -157,8 +159,23 @@ public struct SwiftBuildTool: SwiftCommand {
+                    observabilityScope: customObservabilityScope ?? swiftTool.observabilityScope
+                )
+            },
+        ], uniquingKeysWith: { a, b in
```
this is a bit unclear to me, we are merging a map of 1?

neonichu replied:
We're merging with the providers from `defaultBuildSystemProvider` above, right now this doesn't have an effect, but I wanted to keep it forward compatible.

tomerd commented on Sources/CoreCommands/MinimalBuildTool.swift:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    func didDownloadAllBinaryArtifacts() {
+    }
```
nit: make all of these into one liners `{}`

tomerd reviewed (approved):
very nice

neonichu commented:
@swift-ci please smoke test
